extensions/poetry/main.ts

```diff
@@ -0,0 +1,108 @@
+import { parseDryRun } from "./parse.ts"
+import { red, green, yellow } from "https://deno.land/std@0.150.0/fmt/colors.ts";
+
+class FileBackup {
+  readonly fileName: string
+  readonly fileContent: string | null
+
+  constructor(fileName: string) {
+    this.fileName = fileName
+    this.fileContent = null
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName)
+    } catch (e) { }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent)
+      } else {
+        await Deno.remove(this.fileName)
+      }
+    } catch (e) { }
+  }
+}
+
+// Parse the output of `poetry` subcommands that support the `--dry-run` flag,
+// and submit the packages.
+//
+// Add the `-vvv` flags to get a detailed report of the dependency resolution
+// process. This way, we can track the actual lockfile changes, as specifying
+// `--dry-run` only would not output anything in combination with flags such
+// as `--lock` that do not perform the actual operations.
+async function poetryCheckDryRun(subcommand: string, args: string[]): boolean {
+  // Read and backup the current poetry lockfile contents.
+  const lockfileBackup = new FileBackup('poetry.lock')
+  const manifestBackup = new FileBackup('pyproject.toml')
+
+  await lockfileBackup.backup()
+  await manifestBackup.backup()
+
+  let process = Deno.run({
+    cmd: ['poetry', subcommand, '-vvv', '-n', '--dry-run', ...args.map(s => s.toString())],
+    stdout: 'piped',
+    stderr: 'piped',
+  })
+
+  await process.status()
+  await process.close()
+
+  // If it existed before, restore the previous contents of the lockfile;
+  // otherwise, delete it. This is a workaround to the fact that in poetry
+  // 1.1.x, the `--dry-run` argument does not prevent the lockfile from 
+  // being modified. This is not fixed as of poetry 1.1.14.
+  // Prudently, do the same for the manifest (pyproject.toml).
+  //
+  // See: https://github.com/python-poetry/poetry/pull/5718
+  await lockfileBackup.restoreOrDelete()
+  await manifestBackup.restoreOrDelete()
+
+  const output = new TextDecoder().decode(await process.output())
+  const packages = parseDryRun(output)
+
+  console.log(`Analyzing packages:`)
+  for (const { name, version } of packages) {
+    console.log(`  - ${name} ${version}`)
+  }
+  console.log()
+
+  const jobId = await PhylumApi.analyze("pypi", packages)
+  const jobStatus = await PhylumApi.getJobStatus(jobId)
+
+  if (jobStatus.pass && jobStatus.status === "complete") {
+    console.log(`[${green("phylum")}] All packages pass project thresholds.\n`)
+    return true
+  } else if (jobStatus.pass) {
+    console.warn(`[${yellow("phylum")}] Unknown packages were submitted for analysis, please check again later.\n`)
+    return false
+  } else {
+    console.error(`[${red("phylum")}] The operation caused a threshold failure.\n`)
+    return false
```

Comment:
Do we want the same exit code in both instances?

Reply:
Yes, as we discussed -- I'd go with either passthrough or exit(1) unless there are other compelling reasons or the user could benefit from that, and in that case we need to make sure our exit codes don't clash with Poetry's.

Reply:
I don't remember any discussion regarding this? Generally I think the user very much could benefit from having different error codes here.

And I think the error code you've selected isn't exactly safe from crashing with poetry. Quite the opposite I'd say?

Reply:
Yeah, having `exit(1)` is more of a catch-all in the event that we don't care about specific meaning. If we do care, we need to come up with two numbers neither of which is 1. IIRC you did use some custom exit values at some point; what was your rationale for choosing them?

Reply:
Tested a bunch of `npm` exit codes, noted the approximate ranges, increased by an order of magnitude and picked a number unusual enough to where it wouldn't be easily picked by anyone else trying to do something similar.

Reply:
I was not able to reproduce any error condition in `poetry` that would return anything but `1`. For the sake of consistency, I'd reuse the same exit codes as `npm`.